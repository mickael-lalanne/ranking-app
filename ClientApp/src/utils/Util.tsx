import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
