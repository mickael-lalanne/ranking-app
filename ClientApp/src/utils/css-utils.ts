export const ELEMENT_SIZE: string = '75px';
