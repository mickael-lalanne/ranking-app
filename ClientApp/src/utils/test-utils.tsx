import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
