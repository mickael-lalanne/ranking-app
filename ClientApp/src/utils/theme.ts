import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
