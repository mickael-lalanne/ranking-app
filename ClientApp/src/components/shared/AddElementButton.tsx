import React from 'react';
import { css } from '@emotion/css';
import { element_container_style } from '../shared/ElementPreview';
import AddCircleOutlineIcon from '@mui/icons-material/AddCircleOutline';
import Button from '@mui/material/Button';
import { ELEMENT_SIZE } from '../../utils/css-utils';
import { useAppSelector } from '../../app/hooks';

const AddElementButton = (
    { changeCallback } : {
        changeCallback: (e: React.ChangeEvent<HTMLInputElement>) => void;
    }
) => {
    const loading: boolean = useAppSelector(state => state.application.loading);

    return(
        <Button
            variant="contained"
            component="label"
            className={element_container_style + ' ' + add_elt_btn_style}
            disabled={loading}
        >
            <AddCircleOutlineIcon style={{ width: '40px', height: '40px' }} />
            <input
                type="file"
                accept="image/*"
                multiple
                className={add_element_input_style}
                onChange={changeCallback}
                data-cy="add-element-input"
            ></input>
        </Button>
    );
};

export default AddElementButton;

/**
 * CSS STYLES
 */
const add_elt_btn_style = css({
    margin: '0 !important',
    width: ELEMENT_SIZE,
    height: ELEMENT_SIZE,
    minWidth: '0 !important'
});

const add_element_input_style = css({
    opacity: 0,
    zIndex: -9999,
    height: 0
});
