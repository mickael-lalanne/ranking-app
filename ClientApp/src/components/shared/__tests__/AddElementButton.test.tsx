import * as React from 'react';
import userEvent from '@testing-library/user-event';
import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import AddElementButton from '../AddElementButton';

const changeCallback = jest.fn();

const AddElementButtonTest = () => {
    return <AddElementButton changeCallback={changeCallback} />;
};

describe('Screenshot', () => {
    it('Takes default screenshot', () => {
        const component = renderWithProviders(AddElementButtonTest());

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    beforeEach(() => {
        changeCallback.mockClear();
    })

    it('uploads an element with image', async () => {
        renderWithProviders(AddElementButtonTest());

        const file = new File(['element'], 'elt.png', {type: 'image/png'})
    
        expect(changeCallback).toHaveBeenCalledTimes(0);
        await userEvent.upload(screen.getByTestId('add-element-input'), file);
        // An image is accepted, the change callback should have been called
        expect(changeCallback).toHaveBeenCalledTimes(1);
    });

    it('uploads an element with image', async () => {
        renderWithProviders(AddElementButtonTest());

        const file = new File(['element'], 'elt.mp4', {type: 'video/mp4'})
    
        expect(changeCallback).toHaveBeenCalledTimes(0);
        await userEvent.upload(screen.getByTestId('add-element-input'), file);
        // A video is not accepted, the change callback should not have been called
        expect(changeCallback).toHaveBeenCalledTimes(0);
    });
});
