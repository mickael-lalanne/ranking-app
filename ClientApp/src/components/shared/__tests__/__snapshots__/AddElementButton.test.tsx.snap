// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot Takes default screenshot 1`] = `
<body>
  <div>
    <label
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-13ambje css-1t5b4kh css-sghohy-MuiButtonBase-root-MuiButton-root"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddCircleOutlineIcon"
        focusable="false"
        style="width: 40px; height: 40px;"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
        />
      </svg>
      <input
        accept="image/*"
        class="css-195j79u"
        data-cy="add-element-input"
        data-testid="add-element-input"
        multiple=""
        type="file"
      />
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </label>
  </div>
</body>
`;
